File: internal/dag/graph.go
package dag

import (
	"fmt"
	"strings"
)

// Graph is the adjacency view of a Definition used by the graph algorithms
// (ticket 1.4): cycle detection, topological ordering, reachability, and
// readiness. Loop edges are indexed separately and excluded from adjacency —
// per ADR-003 they participate in expansion (M14), not in readiness or skip
// propagation, and are the only sanctioned cycles.
//
// A Graph is immutable after construction and safe for concurrent use.
type Graph struct {
	def   *Definition
	index map[string]int // step ID → declaration index

	// outNormal / inNormal hold, per step (by declaration index), the
	// indices into def.Edges of its normal out- and in-edges, in edge
	// declaration order.
	outNormal [][]int
	inNormal  [][]int

	// loopEdges holds the indices of the marked loop edges, in declaration
	// order.
	loopEdges []int
}

// NewGraph builds the adjacency view of def. It fails only on definitions
// no graph can be built from — nil, duplicate step IDs, or edges naming
// unknown steps; everything else (including cycles, which Validate rejects)
// is left to the algorithms so their errors carry precise context.
func NewGraph(def *Definition) (*Graph, error) {
	if def == nil {
		return nil, fmt.Errorf("dag: NewGraph called with nil definition")
	}
	g := &Graph{
		def:       def,
		index:     make(map[string]int, len(def.Steps)),
		outNormal: make([][]int, len(def.Steps)),
		inNormal:  make([][]int, len(def.Steps)),
	}
	for i, s := range def.Steps {
		if first, dup := g.index[s.ID]; dup {
			return nil, fmt.Errorf("dag: NewGraph: duplicate step ID %q (steps[%d] and steps[%d])", s.ID, first, i)
		}
		g.index[s.ID] = i
	}
	for i, e := range def.Edges {
		from, ok := g.index[e.From]
		if !ok {
			return nil, fmt.Errorf("dag: NewGraph: edges[%d].from names unknown step %q", i, e.From)
		}
		to, ok := g.index[e.To]
		if !ok {
			return nil, fmt.Errorf("dag: NewGraph: edges[%d].to names unknown step %q", i, e.To)
		}
		if e.IsLoop() {
			g.loopEdges = append(g.loopEdges, i)
			continue
		}
		g.outNormal[from] = append(g.outNormal[from], i)
		g.inNormal[to] = append(g.inNormal[to], i)
	}
	return g, nil
}

// toIdx returns the declaration index of edge i's target step.
func (g *Graph) toIdx(edge int) int { return g.index[g.def.Edges[edge].To] }

// TopoOrder returns the step IDs in a topological order of the normal-edge
// graph (Kahn's algorithm, deterministic: ties break in declaration order).
// It fails if the normal-edge graph has a cycle; Validate is the layer that
// reports cycles with per-edge paths.
func (g *Graph) TopoOrder() ([]string, error) {
	order, ok := g.topoOrder()
	if !ok {
		return nil, fmt.Errorf("dag: TopoOrder: the normal-edge graph has a cycle (run Validate for the offending edges)")
	}
	ids := make([]string, len(order))
	for i, idx := range order {
		ids[i] = g.def.Steps[idx].ID
	}
	return ids, nil
}

// topoOrder is TopoOrder on declaration indices; ok is false when a cycle
// prevents a complete order.
func (g *Graph) topoOrder() (order []int, ok bool) {
	remaining := make([]int, len(g.def.Steps))
	queue := make([]int, 0, len(g.def.Steps))
	for i := range g.def.Steps {
		remaining[i] = len(g.inNormal[i])
		if remaining[i] == 0 {
			queue = append(queue, i)
		}
	}
	order = make([]int, 0, len(g.def.Steps))
	for len(queue) > 0 {
		s := queue[0]
		queue = queue[1:]
		order = append(order, s)
		for _, e := range g.outNormal[s] {
			t := g.toIdx(e)
			if remaining[t]--; remaining[t] == 0 {
				queue = append(queue, t)
			}
		}
	}
	return order, len(order) == len(g.def.Steps)
}

// Ancestors returns the set of step IDs with a non-empty normal-edge path
// to stepID. A step is never its own ancestor (normal-edge cycles are
// rejected by Validate before ancestry is meaningful).
func (g *Graph) Ancestors(stepID string) (map[string]bool, error) {
	start, ok := g.index[stepID]
	if !ok {
		return nil, fmt.Errorf("dag: Ancestors: unknown step %q", stepID)
	}
	ancestors := make(map[string]bool)
	visited := make([]bool, len(g.def.Steps))
	queue := []int{start}
	for len(queue) > 0 {
		s := queue[0]
		queue = queue[1:]
		for _, e := range g.inNormal[s] {
			p := g.index[g.def.Edges[e].From]
			if !visited[p] {
				visited[p] = true
				ancestors[g.def.Steps[p].ID] = true
				queue = append(queue, p)
			}
		}
	}
	return ancestors, nil
}

// reaches reports whether a non-empty normal-edge path leads from step
// index `from` to step index `to`. Starting and ending at the same step
// requires an actual cycle through it, not the trivial empty path.
func (g *Graph) reaches(from, to int) bool {
	visited := make([]bool, len(g.def.Steps))
	queue := []int{from}
	for len(queue) > 0 {
		s := queue[0]
		queue = queue[1:]
		for _, e := range g.outNormal[s] {
			t := g.toIdx(e)
			if t == to {
				return true
			}
			if !visited[t] {
				visited[t] = true
				queue = append(queue, t)
			}
		}
	}
	return false
}

// cycleFinding is one cycle in the normal-edge graph: the edge that closes
// it and the full step path, e.g. ["a", "b", "a"].
type cycleFinding struct {
	edgeIdx int
	path    []string
}

// pathString renders the cycle path for error messages.
func (c cycleFinding) pathString() string { return strings.Join(c.path, " -> ") }

// findCycles locates cycles in the normal-edge graph via iterative
// three-color DFS (iterative so a 10k-step chain cannot overflow the
// stack). It reports one finding per back edge encountered — distinct
// cycles each produce their own finding — deterministically: DFS roots and
// out-edges are visited in declaration order.
func (g *Graph) findCycles() []cycleFinding {
	const (
		white = iota // unvisited
		gray         // on the current DFS path
		black        // fully explored
	)
	color := make([]int8, len(g.def.Steps))
	type frame struct {
		step int
		next int // position in outNormal[step]
	}
	var findings []cycleFinding
	var stack []frame

	for root := range g.def.Steps {
		if color[root] != white {
			continue
		}
		color[root] = gray
		stack = append(stack[:0], frame{step: root})
		for len(stack) > 0 {
			f := &stack[len(stack)-1]
			if f.next == len(g.outNormal[f.step]) {
				color[f.step] = black
				stack = stack[:len(stack)-1]
				continue
			}
			e := g.outNormal[f.step][f.next]
			f.next++
			t := g.toIdx(e)
			switch color[t] {
			case white:
				color[t] = gray
				stack = append(stack, frame{step: t})
			case gray:
				// Back edge: the cycle is the DFS path from t to the top of
				// the stack, closed by e back to t.
				var path []string
				for i := range stack {
					if stack[i].step == t {
						for _, fr := range stack[i:] {
							path = append(path, g.def.Steps[fr.step].ID)
						}
						break
					}
				}
				path = append(path, g.def.Steps[t].ID)
				findings = append(findings, cycleFinding{edgeIdx: e, path: path})
			}
		}
	}
	return findings
}

File: web/lib/graphdef/src/validate/graph.ts
// A pure normal-edge graph over a definition's steps/edges, ported from
// internal/dag/graph.go: the adjacency the graph-semantic checks need
// (findCycles, reaches, Ancestors, loopBody). Loop edges (`type: "loop"`) are
// excluded from readiness adjacency exactly as on the backend. Operates on the
// raw step/edge shapes (after the decode stage has confirmed unique ids and
// resolved endpoints), so it never depends on the typed Definition.

/** A step as the graph reads it: an id and whether it is a loop edge target. */
export interface GraphStep {
  id: string;
}

/** An edge as the graph reads it. `loop` marks an M1 loop edge. */
export interface GraphEdge {
  from: string;
  to: string;
  loop: boolean;
}

/** One cycle finding: the declaration index of the closing edge and the path. */
export interface CycleFinding {
  edgeIdx: number;
  path: string[];
}

export function pathString(c: CycleFinding): string {
  return c.path.join(" -> ");
}

export class Graph {
  readonly steps: GraphStep[];
  readonly edges: GraphEdge[];
  readonly index = new Map<string, number>(); // step id → declaration index
  readonly outNormal: number[][]; // per step: outgoing normal-edge indices
  readonly inNormal: number[][]; // per step: incoming normal-edge indices
  readonly loopEdges: number[] = []; // indices of marked loop edges

  constructor(steps: GraphStep[], edges: GraphEdge[]) {
    this.steps = steps;
    this.edges = edges;
    this.outNormal = steps.map(() => []);
    this.inNormal = steps.map(() => []);
    steps.forEach((s, i) => this.index.set(s.id, i));
    edges.forEach((e, i) => {
      if (e.loop) {
        this.loopEdges.push(i);
        return;
      }
      const from = this.index.get(e.from);
      const to = this.index.get(e.to);
      if (from !== undefined) this.outNormal[from]!.push(i);
      if (to !== undefined) this.inNormal[to]!.push(i);
    });
  }

  private toIdx(edge: number): number {
    return this.index.get(this.edges[edge]!.to)!;
  }

  /**
   * Locate cycles in the normal-edge graph via iterative three-colour DFS in
   * declaration order (dag.Graph.findCycles), one finding per back edge.
   */
  findCycles(): CycleFinding[] {
    const WHITE = 0;
    const GRAY = 1;
    const BLACK = 2;
    const color = new Array<number>(this.steps.length).fill(WHITE);
    const findings: CycleFinding[] = [];

    for (let root = 0; root < this.steps.length; root += 1) {
      if (color[root] !== WHITE) continue;
      color[root] = GRAY;
      const stack: Array<{ step: number; next: number }> = [{ step: root, next: 0 }];
      while (stack.length > 0) {
        const f = stack[stack.length - 1]!;
        const outs = this.outNormal[f.step]!;
        if (f.next === outs.length) {
          color[f.step] = BLACK;
          stack.pop();
          continue;
        }
        const e = outs[f.next]!;
        f.next += 1;
        const t = this.toIdx(e);
        if (color[t] === WHITE) {
          color[t] = GRAY;
          stack.push({ step: t, next: 0 });
        } else if (color[t] === GRAY) {
          const path: string[] = [];
          for (let i = 0; i < stack.length; i += 1) {
            if (stack[i]!.step === t) {
              for (const fr of stack.slice(i)) path.push(this.steps[fr.step]!.id);
              break;
            }
          }
          path.push(this.steps[t]!.id);
          findings.push({ edgeIdx: e, path });
        }
      }
    }
    return findings;
  }

  /** BFS reachability over normal edges (dag.Graph.reaches), by index. */
  reaches(from: number, to: number): boolean {
    const visited = new Array<boolean>(this.steps.length).fill(false);
    const queue = [from];
    while (queue.length > 0) {
      const s = queue.shift()!;
      for (const e of this.outNormal[s]!) {
        const t = this.toIdx(e);
        if (t === to) return true;
        if (!visited[t]) {
          visited[t] = true;
          queue.push(t);
        }
      }
    }
    return false;
  }

  /** The set of step ids strictly upstream of stepID via normal edges (dag.Graph.Ancestors). */
  ancestors(stepID: string): Set<string> {
    const start = this.index.get(stepID);
    const out = new Set<string>();
    if (start === undefined) return out;
    const visited = new Array<boolean>(this.steps.length).fill(false);
    const queue = [start];
    while (queue.length > 0) {
      const s = queue.shift()!;
      for (const e of this.inNormal[s]!) {
        const p = this.index.get(this.edges[e]!.from)!;
        if (!visited[p]) {
          visited[p] = true;
          out.add(this.steps[p]!.id);
          queue.push(p);
        }
      }
    }
    return out;
  }
}

/** Whether the normal-edge graph (all loop edges removed) is well-formed enough to build a Graph. */
export function hasUniqueIds(steps: readonly GraphStep[]): boolean {
  const seen = new Set<string>();
  for (const s of steps) {
    if (seen.has(s.id)) return false;
    seen.add(s.id);
  }
  return true;
}
